### PR TITLE
Fix for #2996

### PR DESCRIPTION
**Pull request by dbaston**
Added implementation of Welzl's algorithm to liblwgeom as a fix to #2996 ([link] (https://trac.osgeo.org/postgis/ticket/2996))

### CODE REVIEW

**Comment by mwtoews** [comment]
Hi @dbaston I was thinking that the C methods should calculate and return the centre point geometry and radius, since these are the results wanted by some folks (e.g. [this Q/A](http://stackoverflow.com/q/31501009/3270260)). This could be a function named something like `ST_MinimumBoundingRadius`, which does everything but buffer (and would not require quadsegs). However I'm aware that returning both a `geometry` and `double` from C is tricky.

This way `ST_MinimumBoundingCircle` could remain as a PL/pgSQL function, and just do a buffer of the centre point geometry and radius from a C function.

**Comment by dbaston** [comment]
@mwtoews Seems logical to me.  Whether implemented in C or PL/pgSQL, is there any way to return a geometry and a double to the user without having to define a custom type, or forcing the user to specify on each function call that the return type is (geometry, double) ?

**Comment by mwtoews** [comment]
In PL/pgSQL this is done with IN and OUT, e.g. written up nicely by @robe2 [here](http://www.postgresonline.com/journal/archives/129-Use-of-OUT-and-INOUT-Parameters.html). But for C, it is not so easy (if I recall).

**Comment by dbaston** [comment]
@mwtoews Updated per your suggestion.  Thanks for the feedback!

**Comment by strk** [comment]
Consider also returning the radius in the M ordinate of the POINT.

**Comment by dbaston** [comment]
OK, I removed the "WIP" qualifier.

Circular geometries are now supported (as they were with the PL/pgsql version)

For a test of cocircular points, I ran `cocircular_points_large_radius_2.data` and `cocircular_points_small_radius_2.data` from [https://github.com/hbf/miniball/tree/master/java/src/test/resources/com/dreizak/miniball/highdim/data].  Results were correct in both cases.

Since all of the action is in liblwgeom and covered by cunit, I don't think SQL-level regression tests beyond `tickets.sql` are needed, but let me know if there's anything specific I should add.

It's probably possible to improve the performance, for example by implementing Welzl's move-to-front heuristic, but at < 300 ms for 1 million random points I'm not sure if it's needed.

File: postgis/postgis.sql.in [hunk 1]
```diff
@@ -3163,6 +3163,17 @@ CREATE OR REPLACE FUNCTION ST_Buffer(geometry,float8,text)
 	   $$
 	LANGUAGE 'sql' IMMUTABLE STRICT;
 
+-- Availability: 2.2.0
+CREATE OR REPLACE FUNCTION ST_MinimumBoundingRadius(geometry, OUT center geometry, OUT radius double precision)
+    AS 'MODULE_PATHNAME', 'ST_MinimumBoundingRadius'
+    LANGUAGE 'c' IMMUTABLE STRICT;
+
+-- Availability: 1.4.0
+CREATE OR REPLACE FUNCTION ST_MinimumBoundingCircle(geometry, segs_per_quarter integer DEFAULT 48)
```
Comment (strk) [inline postgis/postgis.sql.in]: Please do not change the name of the first argument, or upgrades won't work very easily here.

File: liblwgeom/cunit/cu_tester.c [hunk 2]
```diff
@@ -255,18 +259,18 @@ int main(int argc, char *argv[])
...
 cu_noticereporter(const char *fmt, va_list ap)
 {
-  char buf[MAX_CUNIT_MSG_LENGTH+1];
-  vsnprintf (buf, MAX_CUNIT_MSG_LENGTH, fmt, ap);
-  buf[MAX_CUNIT_MSG_LENGTH]='\0';
-  /*fprintf(stderr, "NOTICE: %s\n", buf);*/
+	char buf[MAX_CUNIT_MSG_LENGTH+1];
+	vsnprintf (buf, MAX_CUNIT_MSG_LENGTH, fmt, ap);
+	buf[MAX_CUNIT_MSG_LENGTH]='\0';
+	/*fprintf(stderr, "NOTICE: %s\n", buf);*/
```
Comment (strk) [inline liblwgeom/cunit/cu_tester.c]: Please do not include indentation changes in the patch

File: postgis/postgis.sql.in [hunk 3]
```diff
@@ -3163,6 +3163,17 @@ CREATE OR REPLACE FUNCTION ST_Buffer(geometry,float8,text)
 	   $$
 	LANGUAGE 'sql' IMMUTABLE STRICT;
 
+-- Availability: 2.2.0
```
Comment (strk) [inline postgis/postgis.sql.in]: This will be targetted at 2.3.0, 2.2.0 is already in feature freeze

**Comment by strk** [comment]
I think SQL-level regression tests, focused on minimum bounding circle, are useful. There are a few things that can go wrong between PostgreSQL module and liblwgeom. Often times SRID mismatches or memory handling.

**Comment by dbaston** [comment]
@strk Not sure if I understand about the feature freeze.  This is a bug fix, is it not?

**Comment by strk** [comment]
Sorry I'm missing a link to the ticket containing the bug, can you add it ?
From what I see in the patch there's a new signature being added though, which does not seem to be a fix on itself. I might be missing some context.

**Comment by dbaston** [comment]
Yup, added a link to the top of the PR.  The added signature for `ST_MinimumBoundingRadius` is a byproduct of the fix; among other things, it makes the fix testable.  (You can't really check that all input points are within the output of `ST_MinimumBoundingCircle` for the reasons @mwtoews describes on Trac)

**Comment by strk** [comment]
I see the ticket (https://trac.osgeo.org/postgis/ticket/2996) is also targetting 2.3.0 (changed 3 months ago by @robe2). The RC1 for 2.2.0 is expected today, a bit late to add non-critical fixes.

BTW, the title of that ticket says " ST_MinimumBoundingCircle doesn't always contain all points " which I understand is still the case with the new implementation of ST_MinimumBoundingCircle. So this PR is really an enhancement, not a fix. The enhancement being the new function ST_MinimumBoundingRadius.

**Comment by dbaston** [comment]
In the new implementation the result doesn't always contain all points
because a polygon is an approximation of a circle.  In the current
implementation it does not contain all points because the algorithm is
incorrect.

Whether this is "critical" or "non-critical" is obviously not my call, but
I would think providing functions that return correct results would in
general be a priority.

**Comment by robe2** [comment]
@strk  I had pushed it to 2.3, because I figured Dan wouldn't have it ready by the time we wanted to release 2.2.  I'm okay with pushing it back to 2.2 if it doesn't break @pramsey  thunder :)

**Comment by robe2** [comment]
From @pramsey on IRC - If it passes all the old regression and is more correct, I can see squeezing it in

**Comment by pramsey** [comment]
When I applied patch to trunk and ran regress, I got a test failure in `tickets.sql`, 
```
-- #2996 --
WITH 
  input AS (SELECT 'SRID=4326;POLYGON((26426 65078,26531 65242,26075 65136,26096 65427,26426 65078))'::geometry AS geom),
  mbc   AS (SELECT (mb).center, (mb).radius FROM (SELECT ST_MinimumBoundingRadius(geom) mb FROM input) sq)
SELECT '#2996', radius = ST_Length(ST_LongestLine(geom, center)) FROM input, mbc;
```
With the diff being
```
 #2788|f|Self-intersection|POINT(1 1)
 #2870|Point[GS]
 #2956|t
-#2996|t
+#2996|
 #3172|LINESTRING M (0 0 1,0 0 2)
```

**Comment by dbaston** [comment]
@pramsey Hmm, I'm trying to reproduce.  Is `tickets.sql` the only failure, or does `minimum_bounding_circle.sql` fail too?

**Comment by pramsey** [comment]
My steps:
```
svn co http://svn.osgeo.org/postgis/trunk pgis-test
wget https://github.com/postgis/postgis/pull/67.diff
cd pgis-test
patch -p1 < ../67.diff
./configure && make && make check
```

**Comment by dbaston** [comment]
I think there's a subtle memory issue here.  I'll dig in later this evening.

**Comment by dbaston** [comment]
Or maybe not so subtle.

**Comment by dbaston** [comment]
Closing, as I've totally muddied the history here.  Sorry for the noise.

**Pull request closed without merging**